runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -1110,31 +1476,75 @@ static iree_status_t iree_hal_hip_device_queue_dealloca(
     const iree_hal_semaphore_list_t wait_semaphore_list,
     const iree_hal_semaphore_list_t signal_semaphore_list,
     iree_hal_buffer_t* buffer) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
   iree_hal_hip_device_t* device = iree_hal_hip_device_cast(base_device);
-  IREE_RETURN_IF_ERROR(
-      iree_hal_hip_set_context(device->hip_symbols, device->hip_context));
+  uint64_t queue_affinity_mask =
+      ((iree_hal_queue_affinity_t)1 << device->topology.count);
+  queue_affinity_mask = queue_affinity_mask | (queue_affinity_mask - 1);
+  queue_affinity &= queue_affinity_mask;
+
+  int device_ordinal = iree_math_count_trailing_zeros_u64(queue_affinity);
+
+  if (device_ordinal > device->topology.count) {
+    IREE_TRACE_ZONE_END(z0);
+    return iree_make_status(
+        IREE_STATUS_INVALID_ARGUMENT,
+        "device affinity out of range, maximum device is %" PRIhsz,
+        device->topology.count);
+  }
+
+  queue_affinity = (uint64_t)1 << device_ordinal;
 
+  iree_status_t status = iree_ok_status();
   if (iree_hal_hip_allocator_isa(iree_hal_device_allocator(base_device))) {
-    iree_status_t status = iree_hal_deferred_work_queue_enqueue_dealloc(
-        device->work_queue, wait_semaphore_list, signal_semaphore_list, buffer);
+    iree_hal_hip_device_semaphore_buffer_operation_callback_data_t*
+        callback_data;
+    status = iree_hal_hip_device_make_buffer_callback_data(
+        device, device->host_allocator, queue_affinity, wait_semaphore_list,
+        signal_semaphore_list, buffer,
+        IREE_HAL_HIP_DEVICE_SEMAPHORE_OPERATION_ASYNC_DEALLOC, &callback_data);
+
+    if (iree_status_is_ok(status) && wait_semaphore_list.count == 0) {
+      status = iree_hal_hip_dispatch_thread_add_dispatch(
+          device->topology.devices[device_ordinal].dispatch_thread,
+          &iree_hal_hip_device_perform_buffer_operation_now, callback_data);
+      IREE_TRACE_ZONE_END(z0);
```

Comment:
avoid extra returns - it's not clear if the callback data is cleaned up here (I don't think it is)